File: 09_maximum_slice_problem/max_slice_sum.py
def get_prefix_sum(A):
  store = 0
  ret = []

  for i in A:
    store += i
    ret.append(store)

  return ret

def solution(A):
  P = get_prefix_sum(A)

  n = len(P)
  start = 0
  end = P[0]
  max_slice = end

  for i in xrange(1, n):
    if P[i - 1] < start:
      max_slice = max(max_slice, end - start)
      start = P[i - 1]
      end = P[i]
    else:
      end = max(end, P[i])

  max_slice = max(max_slice, end - start)

  return max_slice

File: 09_maximum_slice_problem/max_slice_sum.cpp
#include <vector>
#include <algorithm>

using namespace std;

vector<long long> prefix_sum(vector<int> &a) {
  int n = a.size();
  vector<long long> ret(n + 1, 0);
  for (int i = 0; i < n; i++) {
    ret[i + 1] = (long long) ret[i] + a[i];
  }
  return ret;
}

int solution(vector<int> &a) {
  int n = a.size();
  vector<long long> psum = prefix_sum(a);
  long long left_end = 0;
  long long ret = psum[1];
  for (int i = 0; i < n; i++) {
    left_end = min(left_end, psum[i]);
    ret = max(ret, psum[i + 1] - left_end);
  }
  return (int) ret;
}
